### DISCUSSION TITLE
Ignoramous asks: is single point tracking of multiple objects possible?

### DISCUSSION THREAD

**Discussion by rackhamlerouge9** [opentrack/opentrack#1543] [Ideas]
With apologies for asking here; I could find no other forum for object tracking, and I understand it's a long shot for opentrack.  I would value a steer of where I should ask the question just as much as I would any answers that apply to opentrack.

I'm enamoured with the [Aerodrums](https://www.aerodrums.com/) product, but they don't sell any software for linux or android.  They use a PS3 eye webcam to track light reflections of drumstick-tip-movement and foot-movement to decide when the movement intersects with a notional 3D drum kit infront of the user.

I'm keen to know:
1. if opentrack could ever be used as a pathway to observe multiple single-point objects, and push it's conclusions to a receiving program?
2. if there're any open source libraries that contribute to single-point object tracking of any form?
3. if there are any forums or other open efforts at which I should take a look?

My thanks for your time in considering all this,
Al

**Comment by sthalik** [comment]
> if opentrack could ever be used as a pathway to observe multiple single-point objects, and push it's conclusions to a receiving program?

You can take parts of `tracker/pt` and use them for that. Particularly, you can cut out https://github.com/opentrack/opentrack/blob/master/tracker-pt/module/point_extractor.h mostly without modifications.

You probably want to replace `pt_frame` in there with something lighter like `cv::Mat3b`, then receive frames from `cv::VideoInput` in a loop.

Point coordinates returned by the extractor are in range of `-0.5 🡒 0.5` for width and height..

> if there're any open source libraries that contribute to single-point object tracking of any form?

Generally you want opencv.

**Reply by rackhamlerouge9** [reply]
Understood, that's super useful - thank you. I'm deeply encouraged!

Please may I ask: is there any obvious way to your mind, other than using different-coloured reflectors, to discriminate left-drumstick from right-drumstick?

**Reply by sthalik** [reply]
Are they composed of multiple points? If so, see `get_d_order` from https://github.com/opentrack/opentrack/blob/master/tracker-pt/point_tracker.cpp. Otherwise you can track them over time (see `find_correspondences_previous` in the same file).

**Reply by rackhamlerouge9** [reply]
Outstanding - they're single points, but I think tracking them over time would be a go-er - thank you so much for this help!!
